### PR TITLE
[Fixes #61] Improve split packages

### LINKED ISSUE #61: Improve SplitPackages
* Refactor the logic for reading packages and classifying them into a separate library that we can use across tools.
* Add support for aggregating packages from multiple folders.
* Enable multiple "traits" on the CSV file and support classifying packages per individual trait.

### PR DESCRIPTION
**Pull request by javiercn**
* Extracted the classification logic into its own class library.
* Added support for multiple source folders with packages.
* Added support for multiple columns in the CSV file.

### CODE REVIEW

**Comment by javiercn** [comment]
@pranavkm @muratg

**Comment by javiercn** [comment]
Ping

**Comment by pranavkm** [comment]
Changes look ok to me (there's a couple of formatting things but code looks ok). That said, is there benefit to doing this when we don't actually have use for it as yet?

**Comment by javiercn** [comment]
@muratg Needs to update the packages.csv file and then we need to update coherence signed to produce the cache using the extra generated project.json

**Comment by javiercn** [comment]
Also we need to integrate this with the push mechanism for pushing packages out. Again @muratg needs to update the packages.csv file

File: src/PackageClassifier/ClassificationEntry.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            Pattern = pattern;
+            Traits = traits;
+        }
+
+        public string Pattern { get; }
+
+        public Trait [] Traits { get; }
+
+        public bool HasPattern(string pattern)
```
Comment (pranavkm) [inline src/PackageClassifier/ClassificationEntry.cs]: Could we remove this for now? https://github.com/aspnet/Coherence/blob/dev/packages/packages.csv doesn't have * (which are problematic regardless).

Comment (pranavkm) [inline src/PackageClassifier/ClassificationEntry.cs]: I meant the `IsLiteralPattern`

Comment (pranavkm) [inline src/PackageClassifier/ClassificationEntry.cs]: At which point we could call this guy `Identifier` or something like that.

File: src/PackageClassifier/Classification.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            while (line != null)
+            {
+                if (!string.IsNullOrWhiteSpace(line))
+                {
+                    lines.Add(line);
+                }
+                line = reader.ReadLine();
+            }
+
+            var traits = lines
```
Comment (pranavkm) [inline src/PackageClassifier/Classification.cs]: Could you add a comment about what the line looks like (similarly one below).

File: src/PackageClassifier/ClassificationEntry.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+
+namespace PackageClassifier
+{
+    public class ClassificationEntry
+    {
+        public ClassificationEntry(string pattern, Trait [] traits)
```
Comment (pranavkm) [inline src/PackageClassifier/ClassificationEntry.cs]: Formatting  - `Trait[] traits` (couple other places with the box brackets having a space before)

File: src/PackageClassifier/ClassifiedPackages.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public class ClassifiedPackages
+    {
+        public ClassifiedPackages(Trait trait, IList<PackageInformation> packages)
+        {
+            Trait = trait;
+            Packages = packages;
+        }
+
+        public Trait Trait { get; private set; }
```
Comment (pranavkm) [inline src/PackageClassifier/ClassifiedPackages.cs]: `public Trait Trait { get; } `

File: src/PackageClassifier/Classifier.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
```
Comment (pranavkm) [inline src/PackageClassifier/Classifier.cs]: Remove and sort?

File: src/PackageClassifier/PackageMapping.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+            Packages = packages;
+        }
+
+        public ClassificationEntry Entry { get; }
+
+        public PackageInformation[] Packages { get; }
+
+        public bool HasPackage(PackageInformation package)
+        {
+            return Packages.Any(p => string.Equals(p.FullPath, package.FullPath));
```
Comment (pranavkm) [inline src/PackageClassifier/PackageMapping.cs]: `StringComparison.Ordinal`?

File: src/PackageClassifier/Classifier.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        public IList<PackageMapping> PackageMappings { get; } = new List<PackageMapping>();
+
+        public IList<string> Diagnostics { get; } = new List<string>();
+
+        public PackageSourcesCache Cache { get; }
+
+        public Classification Classification { get; }
+
+        public ClassificationResult GetClassificationFor(string trait)
```
Comment (pranavkm) [inline src/PackageClassifier/Classifier.cs]: `GetClassification`?

File: src/PackageClassifier/project.json [hunk 8]
```diff
@@ -0,0 +1,32 @@
+{
+  "version": "1.0.1-*",
+  "description": "Classifies a set of packages from several directories based on information from a csv file",
+  "authors": [ "Microsoft" ],
+  "dependencies": {
+    "Microsoft.Extensions.Logging": "1.0.0-*",
+    "Microsoft.Extensions.Logging.Console": "1.0.0-*",
+    "Newtonsoft.Json": "8.0.2",
```
Comment (pranavkm) [inline src/PackageClassifier/project.json]: 9.0.1-beta

File: src/SplitPackages/Program.cs [hunk 9]
```diff
@@ -120,35 +127,47 @@ private int Execute()
                     return Error;
                 }
 
+                if (_debug.HasValue())
+                {
+                    Console.WriteLine($"PROCESS ID: {System.Diagnostics.Process.GetCurrentProcess().Id}");
+                    Console.Out.Flush();
```
Comment (pranavkm) [inline src/SplitPackages/Program.cs]: Shouldn't need this.

File: src/SplitPackages/Program.cs [hunk 10]
```diff
@@ -120,35 +127,47 @@ private int Execute()
...
+                    while (!System.Diagnostics.Debugger.IsAttached)
```
Comment (pranavkm) [inline src/SplitPackages/Program.cs]: `using System.Diagnostics;`

**Comment by pranavkm** [comment]
:watch: for the pattern changes

**Comment by muratg** [comment]
I think we can check this in and make the packages.csv changes later.

**Comment by javiercn** [comment]
Updated @pranavkm

File: src/PackageClassifier/Classification.cs (hunk 2 above)
Comment (pranavkm) [inline src/PackageClassifier/Classification.cs]: Comment?

**Comment by pranavkm** [comment]
:shipit:

**Comment by muratg** [comment]
:shipit:

**Pull request merged**